### ISSUE TITLE
BUG: scalars missing several methods for array api compat

### ISSUE THREAD

**Issue by Illviljan** [numpy/numpy#27305]
### Describe the issue:

I keep getting stuck trying to get tests in https://github.com/data-apis/array-api-tests to pass and the final errors are often due to arrays having turned into scalars and the tests not expecting that, (https://data-apis.org/array-api/draft/API_specification/array_object.html)

Reading #26850 it seems the intention is that the scalars should support the array api spec?

### Reproduce the code example:

```python
import numpy as np
import array_api_strict as xps


# Some examples:
xps.mean(xps.asarray(4, dtype=xps.float32)).__iadd__(1)
np.mean(np.asarray(4, dtype=np.float32)).__iadd__(1) # AttributeError: 'numpy.float32' object has no attribute '__iadd__'

xps.mean(xps.asarray(4, dtype=xps.float32)).__complex__()
np.mean(np.asarray(4, dtype=np.float32)).__complex__() # AttributeError: 'numpy.float32' object has no attribute '__complex__'
```

### Python and NumPy Versions:

import sys, numpy; print(numpy.__version__); print(sys.version)
2.1.0
3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:42:31) [MSC v.1937 64 bit (AMD64)]

**Comment by ngoldbaum** [comment]
So one issue with the in-place operators is that scalars are immutable. Maybe we could implement them but make them return a copy? But maybe a in-place operation returning a copy is confusing?

Implementing `__complex__` makes sense. I guess the reason it's missing is there isn't a `PyNumberMethod` slot for it as far as I can see: https://docs.python.org/3/c-api/typeobj.html. You'd probably need to "manually" define a function named `__complex__`.

**Comment by ngoldbaum** [comment]
This is the current behavior:

```
In [6]: a = np.int64(3)

In [7]: id(a)
Out[7]: 2199088643840

In [8]: a += 3

In [9]: id(a)
Out[9]: 2199088641120
```

I think implementing `__iadd__` and making it return a copy doens't actually change this behavior. So we should do that.

I added a 2.2.0 release candidate. 

I think `__iadd__` and `__ior__` are probably the easiest since there are `PyNumberMethod` slots for them. 

For `__complex__` someone would need to define a python function (in C) named `__complex__` that does the conversion to a PyComplex.

**Comment by Illviljan** [comment]
Setting up [array-api-tests](https://github.com/data-apis/array-api-tests) to run with scalars will probably be helpful as well. 
It might smoke out more missing methods.

**Comment by Ishankoradia** [comment]
@ngoldbaum have we decided to do this ?
I wanted to give it a shot.

**Comment by ngoldbaum** [comment]
@Ishankoradia go ahead. Having a milestone means we're planning to do it. We also don't claim issues, just go ahead and work on it.

**Comment by Ishankoradia** [comment]
Gotcha !!

@ngoldbaum I am looking at the `core/__init__.pyi`, i can see that for method `__iadd__` we only accept NDArray. I am guessing my first step would be to add a method overload to accept scalar. Am i in the right direction ?

**Comment by ngoldbaum** [comment]
No, in order to implement these functions you'll need to modify the C internals of NumPy.

Here is where the `PyNumberMethods` struct is set up for all of the NumPy scalar types:

https://github.com/numpy/numpy/blob/f37a4f32887595e710ecb948c811e4b092b20be5/numpy/_core/src/umath/scalarmath.c.src#L1974-L1997

This is inside a file that is written in NumPy's custom templating language used for codegen internally.

You can see that none of the `inplace` methods listed in the [CPython docs](https://docs.python.org/3/c-api/typeobj.html#c.PyNumberMethods) are implemented. I think we need to implement all of them, not just `__add__` and `__or__`. You'd also need to add tests and I'd also double check my analysis above that defining in-place operators that return copies is actually OK. Do other array libraries do that?

For `__complex__`, as noted above, you'd need to add an entry and implementation for `__complex__` to the `PyMethodDef` array for the scalar types: https://github.com/numpy/numpy/blob/f37a4f32887595e710ecb948c811e4b092b20be5/numpy/_core/src/multiarray/scalartypes.c.src#L2586.

If you've never touched the CPython C API before this is probably a big project, although tbh it would be a decent way to learn about the C internals of NumPy or how to work with Python C extensions using the C API directly.

**Comment by Ishankoradia** [comment]
Thanks a ton @ngoldbaum , this helps a lot. 
I think I can do it. Although i have limited knowledge of Python C extensions , but like you said what better to learn this & numpy C internals.

**Comment by ngoldbaum** [comment]
I found going through https://docs.python.org/3/extending/extending.html and https://llllllllll.github.io/c-extension-tutorial/ helped immensely to understand this stuff better.

**Comment by Ishankoradia** [comment]
@ngoldbaum I have spent a lot of time reading the material you shared. I have good understanding of how c extensions work. 

I was looking at this template file you pointed out `scalarmath.c.src`. Its very interesting. I think all `@name@` placeholders are replaced by the correct dtypes. And i also see their corresponding method implementation in the compiled file `scalarmath.c`. But i can figure out where is the source code for those functions (eg. `@name@_add` ) coming from. Could you point me to that file ?

**Comment by ngoldbaum** [comment]
Hi, sorry for taking a few days to respond, I've been on vacation.

The `@name@` part of the template system numpy uses for codegen. All files with .c.src extensions use this template system.

The templating for the block of code I linked to is set up immediately above that code:

https://github.com/numpy/numpy/blob/f37a4f32887595e710ecb948c811e4b092b20be5/numpy/_core/src/umath/scalarmath.c.src#L1968-L1973

Here, the template system is saying to replace `@name@` with each of the names in the comma-separated list, one for each scalar type.

In order to implement `__iadd__`, `__ior__`, and the rest of the in-place operators, you're going to need to define new templates that define implementations for the these operators (or maybe you can just re-use the existing implementations for the non-in-place operators? not sure) then add new entries to the table I linked to above for all the in-place operators like `nb_inplace_add`.

**Comment by Ishankoradia** [comment]
@ngoldbaum no problem. Thanks for getting back. (Also, hope you had a great vacation and got good time to recharge)

So my first attempt was to use the same table. I tried to add a new entry there for `.nb_inplace_add` like this
<img width="565" alt="Screenshot 2024-10-07 at 21 53 36" src="https://github.com/user-attachments/assets/26bb333f-20ad-4411-bb92-2b35cc3caf30">

But the mistake I did was, I implemented a method `@name@_ctype_inplace_add` instead `@name@_inplace_add`. I assumed `ctype` is also some kind of a prefix stub that is needed for the methods because i see it everywhere.

I have added this now and i see the compiled file `scalarmath.c` that has this dummy implementation.
<img width="565" alt="Screenshot 2024-10-07 at 22 11 48" src="https://github.com/user-attachments/assets/197f498b-6041-4d33-af59-c234bb695392">


1. What is that `ctype` prefix ?
2. Right now I have just copied the dummy implementation from `_ctype_add`. I guess i will have to break it down once i start implementing for each data types right based on how the logic looks like ?
3. How are inputs to these methods handled here ? Can i assume in inplace operation that the first input will be `self` ?

If you can clear these up , it would be great. Sorry for the back & forth. Thank you for all the help in this one.

**Comment by Ishankoradia** [comment]
> > What is that ctype prefix ?
> 
> > Right now I have just copied the dummy implementation from _ctype_add. I guess i will have to break it down once i start implementing for each data types right based on how the logic looks like ?
> 
> The ctype functions are used to define the "c-level" version of the operation. There's also a "python-level" version of the operation that is defined using a second level of templating to define a function named e.g. `float_add`. See the template function defined starting at line 1175 in scalarmath.c.src. I guess to get this to work, you'll also need to extend this to generate in-place python wrappers, along with the C-level wrappers. Although that said, maybe you can just re-use the existing wrappers that are already defined, so just set `nb_inplace_add` to e.g. `(binaryfunc)@name@_add`.
> 
> If you look up at my analysis when I originally opened the issue - numpy scalars already let you *use* the in-place operators, they just return a copy, so maybe in principle you can just re-use the existing implementations that are getting used already, but now by explicitly using the slot so people can call the dunder methods directly from Python.
> 
> > How are inputs to these methods handled here ? Can i assume in inplace operation that the first input will be self ?
> 
> I think that's right, although instead of asking me and then waiting for a response, you should try experimenting to see for yourself. I would need to poke around with a C debugger to figure that out.
> 
> There are `spin gdb` and `spin lldb` commands. You'll also need to make sure you're building NumPy with debug symbols. There's no need to use a debug build of CPython unless you need to step through CPython, which is not needed for most things.
> 
> If you don't like debuggers, printf debugging also works :)

Understood. Thank you Nathan !!
This helps a lot. I will get back in 3-4 days hopefully with something implemented.

**Comment by mhvk** [comment]
Sorry for a late side comment, but reading this are we sure that this https://github.com/numpy/numpy/issues/27305#issuecomment-2383799774 has been followed up:
> You can see that none of the inplace methods listed in the [CPython docs](https://docs.python.org/3/c-api/typeobj.html#c.PyNumberMethods) are implemented. I think we need to implement all of them, not just __iadd__ and __ior__. You'd also need to add tests and I'd also double check my analysis above that defining in-place operators that return copies is actually OK. Do other array libraries do that?

Do other array libraries in fact do that for immutable scalars? I ask because certainly `python` ints do *not* implement `__iadd__` - within python at least those methods should not be used directly, but tested through simply doing `a += b` (which goes through `a.__add__(b)` if there is no `__iadd__`).

**Comment by ngoldbaum** [comment]
This is related to #27137

**Comment by jorenham** [comment]
Note that `builtins.float`, `int`, and `bool`, also don't implement `__complex__`, because there simply is no need for it:

Constructing a `builtins.complex` from an object does not require a `__complex__` to be implemented, as long as it implements `__float__` or `__index__`. This can be seen in the [typeshed stubs for `builtins.complex`](https://github.com/python/typeshed/blob/94ffa6e3f512b491252a1d0674fef8a3f283534d/stdlib/builtins.pyi#L432-L469), for example. 

All of NumPy's floats, integers, and `bool`, implement the `__float__` method, and are therefore already convertible to `builtins.complex`. So adding a `__complex__` to them won't help with that.

The only use-case for `__complex__` I can think of is for custom complex-like types, like the ones in `complexfloating`. Otherwise, there you wouldn't be able to convert them to a `builtins.complex` without them losing the imaginary part, since `__float__` will ignore that. 

But maybe there there's some other reason for why the Array API requires `__complex__`?

**Comment by Illviljan** [comment]
A package implementing the array standard will always return array objects, 0d arrays rather than scalars. 

Numpy sometimes returns scalars. 
Some would argue numpy is not conforming to the array standard, numpy argues that scalars will ducktype as a valid array api object (requiring `__complex__`).

**Comment by jorenham** [comment]
Everything in the array api standard serves a specific purpose. I'm trying to figure out if that purpose also applies to the context at hand. Because it is possible that this scenario was overlooked while designing the spec. 
So I don't think it's a good idea to blindly follow the array api (or any other specification for that matter) purely for the sake of "conforming" to it.
If there's a red traffic light in the middle of an empty desert, I also wouldn't stop and wait for it to turn green ;)

**Comment by seberg** [comment]
Discussing scalars in NumPy is quite frankly way off-topic and irrelevant, if you wish to discuss it there are some issues that bring up the necessary nuances that are lost here anyway. 

If we start that here, anyone with a bit of C knowledge could implement and review the relevant addition here (before 2.4), and it is important enough to review.
This topped nobodies priorities, maybe it should have considering it how it easy it is, but it didn't.  I'd certainly review a PR.

**Comment by seberg** [comment]
I tend to agree with Joren, we could add it just for the heck of it, since it shouldn't be problematic either. But it would be nice to understand better why the `array-api-tests` tell you about this causes problems; because to me it doesn't really seem nicer to do something Python doesn't do.
For all I know we can just modify those tests to say what is correct: `__complex__` needs to work for complex (and really testing `complex(...)` may be good enough -- even if that is less strict with strings).

(That said, there are real other holes that should be fixed, just complex is a bit unclear.)

**Comment by jorenham** [comment]
> since it shouldn't be problematic either.

It might be for static typing though. Because using `SupportsComplex` is sometimes used to workaround the fact that type-checkers accept `float` when you annotate something to accept `complex` (which is a bit of a lie, because at runtime, `float` isn't a subclass of `complex`).

**Comment by Illviljan** [comment]
> But it would be nice to understand better why the array-api-tests tell you about this causes problems; because to me it doesn't really seem nicer to do something Python doesn't do.

I feel we are going in circles. Here's a summary how I remember it:

* Array Api standard optimistically wants conforming array libraries to _always_ return array objects. numpy never did pre-2.0 maybe it could change post-2.0? [NEP 47 argues scalars are largely a design mistake.](https://numpy.org/neps/nep-0047-array-api-standard.html#indexing)
* numpy decided not to change, "scalars **are** ducktyping 0-D arrays".  If scalars are missing anything is that just an oversight?
* Several oversights missed and fixed, examples #27137, #26889. 
* `array-api-tests` starts development assuming array ops _always_ returns arrays, their Hypothesis-based testing leads to a lot of edge case testing.
* This issue examplifies another oversight by not consistently returning arrays; [Array object must have an __complex__ method](https://data-apis.org/array-api/draft/API_specification/generated/array_api.array.__complex__.html#array_api.array.__complex__). Just slap on a `__complex__` like before? 
* Slapping on `__complex__` makes little sense on scalars, python doesn't do that. So numpy scalars **aren't** ducktyping 0-D-arrays then?

I noticed these edge cases because I was testing an array-api compliant object that wraps _another_ array-api compliant objects and the quirks of each implementation can be tricky to deal with. 
Was a while ago I thought about this issue in detail but I think there was a workaround. (Isn't there always?)

My big picture question is, do we agree that it would be odd if numpy struggles to comply with the array-api standard?
If so, scalars maybe needs some strange methods just for the sake of being strictly array-api compliant?

**Comment by jorenham** [comment]
Maybe it's just me, but reasons like "because the array API says so" or "others also do it" just don't convince me.

So I'll ask again: do you have a concrete real-life code example that demonstrates the need for adding  `__complex__` to all `numpy.number` scalar types?
